# Lab3 Police Shootings

# Introduction
We are looking at police shootings and possible correlations to race
- What corelates high cause of death in police shootings?
    - Does race change how a police officer deals with a situation
        - Is there any correlation between race of someone killed by police and how they were killed
    - Does gender change how a police officer deals with a situation
        - Is there any correlation between gender of someone killed by police and how they were killed
    - Are people who are armed more likely to be shot by police
        - Is there any correlation between whether or someone who is armed who is killed by police and how they were killed


## Dataset Preparation
#### Source data:
- Data on police shootings
- Data on police violence
- Data on different states in the US

#### Preparing
- In excel
    - I cleaned up headers to take out apostrophes in csv files through excel
    - Then I resaved the file with different encoding, so it could be opened with Jupiter
- Using Python
    - Loaded csv's into data frames
    - I replaced many of the string values I wanted to use with numerical values
    - Then anything that wasn't already turned into integers then was
    - After running correlations with data and finding low r values, I also decided to try changing unarmed status to just whether the victim was allegedly armed. And changing the cause of death to whether the victim was shot. 


In [9]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import random as rnd
import math

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# machine learning
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier

#(yes i know i probably dont need most of this...)

In [10]:
import csv
import math
import statistics as stats
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

kbp = pd.read_csv('KilledByPolice2.csv') #read in file to dataframe
kbp[:3]

#create original copy
kbpOG=kbp

In [11]:
pvc = pd.read_csv('PoliceViolence_city2.csv') #read in file to dataframe
pvc[:3]

#create original copy
pvcOG=pvc

In [12]:
#note this data has no header
states = pd.read_csv('states2.csv') #read in file to dataframe
states[:3]

#create original copy
statesOG=states

In [13]:

r=0
#- Is the % black killed by police of people killed by police porportional to % black people in the community 
totalPop= pvc['Total']
blackPop= pvc['Black']
blackKilled= pvc['Black People Killed by Police (2013-2015)']

#print(totalPop, blackPop, blackKilled)


r = np.corrcoef((pvc['Total']/pvc['Black']), pvc['Black People Killed by Police (2013-2015)'])[0, 1]
print("Correlation in %black in community over black killed by police is ",r)

#- does race corelate with cause of death for police killings

#kbp[3] = kbp[3].map({'Female': 1, 'Male': 0, 'Unknown': 3, 'Transgender': 3})

#change genders to numbers
kbp['Victims gender'][228]='Unknown' #fill in the one blank cell
kbp['Victims gender'] = kbp['Victims gender'].replace('Male', 0)
kbp['Victims gender'] = kbp['Victims gender'].replace('Female', 1)
kbp['Victims gender'] = kbp['Victims gender'].replace('Unknown', 3)
kbp['Victims gender'] = kbp['Victims gender'].replace('Transgender', 3)

#Change race to numbers
kbp['Victims race'] = kbp['Victims race'].replace('Asian', 0)
kbp['Victims race'] = kbp['Victims race'].replace('Black', 1)
kbp['Victims race'] = kbp['Victims race'].replace('Hispanic', 2)
kbp['Victims race'] = kbp['Victims race'].replace('Native American', 3)
kbp['Victims race'] = kbp['Victims race'].replace('Pacific Islander', 4)
kbp['Victims race'] = kbp['Victims race'].replace('Unknown race', 5)
kbp['Victims race'] = kbp['Victims race'].replace('White', 6)


#Change cause of death to numbers
kbp['Cause of death'] = kbp['Cause of death'].replace('Asphyxiated', 0)
kbp['Cause of death'] = kbp['Cause of death'].replace('Baton, Pepper Spray, Physical Restraint', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Bean bag', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Bean bag, taser', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Beaten', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Beating', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Death in custody', 2)
kbp['Cause of death'] = kbp['Cause of death'].replace('Drowned', 3)
kbp['Cause of death'] = kbp['Cause of death'].replace('Drowning', 3)
kbp['Cause of death'] = kbp['Cause of death'].replace('Drug Overdose', 4)
kbp['Cause of death'] = kbp['Cause of death'].replace('Fall to death', 5)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot Stabbed', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot, Taser', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot, Taser, Pepper spray', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot, Vehicle', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Hanging', 7)
kbp['Cause of death'] = kbp['Cause of death'].replace('Medical Emergency', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Medical emergency, unknown', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Medical emergencys', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Neglect', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Negligence, failure to call paramedics when subject could not breathe due to asthma', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Other', 9)
kbp['Cause of death'] = kbp['Cause of death'].replace('Pepper Spray', 10)
kbp['Cause of death'] = kbp['Cause of death'].replace('Pepper sprayed, Taser, beaten', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Physical restraint', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Smoke inhilation', 11)
kbp['Cause of death'] = kbp['Cause of death'].replace('Struck by vehicle', 12)
kbp['Cause of death'] = kbp['Cause of death'].replace('Suicide', 7)
kbp['Cause of death'] = kbp['Cause of death'].replace('Taser', 13)
kbp['Cause of death'] = kbp['Cause of death'].replace('Taser, Beaten', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Taser, Medical Emergency', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Taser, Physical Restraint', 13)
kbp['Cause of death'] = kbp['Cause of death'].replace('Taser/Pepper spray/beaten', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Tasered', 13)
kbp['Cause of death'] = kbp['Cause of death'].replace('Undetermined', 14)
kbp['Cause of death'] = kbp['Cause of death'].replace('unknown', 14)
kbp['Cause of death'] = kbp['Cause of death'].replace('Vehicle', 12)

kbp['Cause of death'] = kbp['Cause of death'].replace('Pepper spray', 10)
kbp['Cause of death'] = kbp['Cause of death'].replace('Physical restraint', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot, Taser', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot, Stabbed', 6)
kbp['Cause of death'] = kbp['Cause of death'].replace('Asphyxiation', 0)
kbp['Cause of death'] = kbp['Cause of death'].replace('Taser, Medical emergency', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Medical emergency', 8)
kbp['Cause of death'] = kbp['Cause of death'].replace('Physical Restraint', 1)
kbp['Cause of death'] = kbp['Cause of death'].replace('Unknown', 14)
kbp['Cause of death'] = kbp['Cause of death'].replace('Death in Custody', 2)
kbp['Cause of death'] = kbp['Cause of death'].replace('Gunshot, Stabbed', 2)

#Change unarmed status to numbers
kbp['Unarmed'] = kbp['Unarmed'].replace('Allegedly Armed', 0)
kbp['Unarmed'] = kbp['Unarmed'].replace('Unarmed', 1)
kbp['Unarmed'] = kbp['Unarmed'].replace('Unclear', 2)
kbp['Unarmed'] = kbp['Unarmed'].replace('Vehicle', 3)


print("done changing to numbers")


Correlation in %black in community over black killed by police is  -0.165486103483
done changing to numbers


C:\Users\maxsk\OneDrive\Documents\Programs\ANACONDA\lib\site-packages\ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


In [14]:
print("start")
#Check to see if a string is castable to an int
#and cast to ints if commented right 
#def isNumber(s):
#    try:
#        s=int(s))
#        #print("s=",s)
#        return s
#    except ValueError:
#        return -1
    

    
#Cast all the all the cells in a 1d list or data frame passed to it to int if they are able to be 
#pass to is (dataframe collum, length stuff you want to go through)
def cellsStrNumToInt(data):
    i=0
    #x=0
    n=len(data)
    while(i<n):    
        
        
        #check to make sure everything is castable before hand and then cast to ints
        #print(i)
        #s = isNumber(data[i])
        #if(s != -1):
        #    x=x+1
        #    data[i]=s
        #else:
        #    print(data[i])
        #    print(i)
        
        #print every 100th line to show that the code is still running and show how much longer its taking
        #if(i%100==0):
        #    print(i)
        data[i]=int(data[i])
        i=i+1

#call method above to cast to ints
#cellsStrNumToInt(kbp['Cause of death'])
#print("done with cause of death")

#cellsStrNumToInt(kbp['Victims race'])
#print("done with victims race")

#cellsStrNumToInt(kbp['Unarmed'])
#print("done with Unarmed")

#print("done cellsStrNumToInt for all")


#changes all the data thats not the value passed to it, to -1 
def anythingBut(data, value):
    i=0
    #x=0
    n=len(data)
    while(i<n):    
        if(data[i] != value):
            data[i]=-1
        i=i+1

anythingBut(kbp['Cause of death'], 6) 
print('cause of death is done')
anythingBut(kbp['Unarmed'], 0)  

print("done")
kbp.head()
#kbp.info()
#r = np.corrcoef(kbp['Victims race'], kbp['Cause of death'])[0, 1]
#print("its",r)

start


C:\Users\maxsk\OneDrive\Documents\Programs\ANACONDA\lib\site-packages\ipykernel_launcher.py:59: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


cause of death is done
done


Victims name Victims age  Victims gender  Victims race  \
0  Keith Lamont Scott     Unknown               0             1   
1    Terence Crutcher          40               0             1   
2         Dahir Aidan          22               0             1   
3      Nicholas Glenn          25               0             1   
4     Thomas Mone III          46               0             6   

                              URL of image of victim  \
0                                                NaN   
1  http://assets.nydailynews.com/polopoly_fs/1.27...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

  Date of injury resulting in death (month/day/year)  \
0                                          9/20/2016   
1                                          9/17/2016   
2                                          9/17/2016   
3                                          9/16/2016   
4                                          9/16/2016   

  Location of injury (address) Location of death (city)  \
0                          NaN                Charlotte   
1      36th St N and Lewis Ave                    Tulsa   
2             4101 Division St                St. Cloud   
3                100 S 48th St             Philadelphia   
4                 Boo Boo Blvd                Hollister   

  Location of death (state)  Location of death (zip code)  \
0                        NC                           NaN   
1                        OK                       74110.0   
2                        MN                       56301.0   
3                        PN                       19139.0   
4                        MO                       65672.0   

  Location of death (county)  \
0                Mecklenburg   
1                      Tulsa   
2                    Stearns   
3               Philadelphia   
4                      Taney   

                        Agency responsible for death  Cause of death  \
0            Charlotte-Mecklenburg Police Department               6   
1                            Tulsa Police Department               6   
2                             Avon Police Department               6   
3                     Philadelphia Police Department               6   
4  Missouri State Highway Patrol Taney County She...               6   

  A brief description of the circumstances surrounding the death  \
0  Keith was shot by an undercover officer while ...               
1  Terence was killed as he leaned on his vehicle...               
2  Adan stabbed nine people at a shopping mall wh...               
3  Glenn was fatally shot during a shootout with ...               
4  Mone was killed in an exchange of gunfire with...               

  Official disposition of death (justified or other) Criminal Charges?  \
0                              Pending investigation  No Known Charges   
1                              Pending investigation  No Known Charges   
2                              Pending investigation  No Known Charges   
3                              Pending investigation  No Known Charges   
4                              Pending investigation  No Known Charges   

  Link to news article or photo of official document  \
0  http://www.cbsnews.com/news/person-killed-in-o...   
1  http://www.cnn.com/2016/09/20/us/oklahoma-tuls...   
2  http://www.sctimes.com/story/news/local/2016/0...   
3  http://www.philly.com/philly/news/20160918_Wha...   
4  http://bransontrilakesnews.com/news_free/artic...   

  Symptoms of mental illness?  Unarmed  
0                          No       -1  
1                          No       -1  
2                     Unknown        0  
3                          No        0  
4                          No        0

## Data Modeling
- Mathematical/statistical model used:
    - I used Pearson Corelation Coefficients that I ended up coding in myself instead of using the package because I has having problems debugging code and mine was more straightforward and easier to diagnose problems in and uses the Pearson Corelation Coefficient formula to find the r values
        - I used Pearson corelation coefficients because I wanted to be able to find if there was any corelation in the data and Pearson corelation coefficients is a terrific way to do that
- I found the corelation in race, gender, and armed status, vs cause of death 
    - I broke up armed status so that it was whether the victim was armed 
        - It had originally been a variety of things such as unknown, but I had done corelations with that and it had been a pretty low r 
    - I also broke up cause of death to whether they were shot in
        - I had originally broken it into almost all the groups that the data gave me which was a lot, but again the corelation was low so I grouped it 



In [15]:
def sumOfData(data):
    i=0
    total=0
    n=len(data)
    while(i<n):
        total = total +data[i]
        
        i=i+1
    return total


def sumOfSqr(data):
    i=0
    total=0
    n=len(data)
    while(i<n):
        total = total + (data[i]*data[i])
        
        i=i+1
    return total

def sumOfProductOfPairs(data1, data2):
    i=0
    total=0
    n=len(data1)
    if(n!=len(data2)):
        raise ValueError('Data is of different lengths aaah')
                         
    while(i<n):
        total = total + (data1[i]*data2[i])
        i=i+1
    return total

def pearson(x,y):
    n=len(x)
    if(n!=len(y)):
        raise ValueError('x,y is of different lengths aaah')
    
    sumOfx=sumOfData(x) 
    sumOfy=sumOfData(y)
                         
    a=(n * sumOfProductOfPairs(x,y)) - (sumOfx * sumOfy)
    b=((n * sumOfSqr(x)) - (sumOfx * sumOfx))
    c=((n * sumOfSqr(y)) - (sumOfy * sumOfy))                  
    
    r=a/(math.sqrt(b*c))
    return r

r = pearson(kbp['Victims race'], kbp['Cause of death'])
print("race vs cause of death rval: ",r)
r = np.corrcoef(kbp['Victims race'], kbp['Cause of death'], True)[0, 1]
print("race vs cause of death rval2: ",r)


r = pearson(kbp['Victims gender'], kbp['Cause of death'])
print("gender vs cause of death rval: ",r)
r = np.corrcoef(kbp['Victims gender'], kbp['Cause of death'], True)[0, 1]
print("gender vs cause of death rval2: ",r)


r = pearson(kbp['Unarmed'], kbp['Cause of death'])
print("armed vs cause of death rval: ",r)
r = np.corrcoef(kbp['Unarmed'], kbp['Cause of death'], True)[0, 1]
print("armed vs cause of death rval2: ",r)


race vs cause of death rval:  0.050165506744
race vs cause of death rval2:  0.050165506744
gender vs cause of death rval:  -0.0760723610656
gender vs cause of death rval2:  -0.0760723610656
armed vs cause of death rval:  0.421091797032
armed vs cause of death rval2:  0.421091797032


In [16]:


#check to make sure everythings not strings
#i=0
#while(i<=4320):    
#    i=i+1
#    if(i%100==0):
#            print(i)
#    kbp['Victims race'][i] = kbp['Victims race'][i] +0
#    kbp['Cause of death'][i] = kbp['Cause of death'][i] +0
    
#kbp.head()
#kbp.info()

#kbp['Victims race'].length = 5
#kbp['Cause of death'].length = 5

#print(kbp['Victims race'])
#print(kbp['Cause of death'])
r = np.corrcoef(kbp['Victims race'], kbp['Cause of death'], True)[0, 1]
print("its",r)

its 0.050165506744


## Data Analysis and Conclusion
- I found the r values for the following correlations: 
    - armed vs cause of death rval: 0.42
    - gender vs cause of death: -0.076
    - race vs cause of death: 0.05
- I found that race or gender does not have a corelation with whether or not someone is shot or killed in another manner if they are killed by the police.
   - This was a bit surprising to me, as I would have thought everything would have correlated a lot more, I would have expected an armed person to be more likely to be shot than killed by something else such as pepper spray or taser 
   - I also expected race to correlate higher, because we always hear about police shootings and race through the media, although it should be noted that this could be that the number of killings is higher, but the method is the same. 

Farther study:
- I would like to have done more in terms of the number of killings instead of just the way they were killed
- I also would have looked more into the percentages of people killed instead of correlation everything


# Acknowledgements
Tracey helped me with some of the code
Louis Lafreniere helped me debug